### PR TITLE
Add support for Python3 in vmware_guest_find

### LINKED ISSUE #25984: vmware_guest_find is not py3 compatible
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME
vmware_guest_find

##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.4.0 (VCENTER_CLOUD_PROVIDER 195fadb2fe) last updated 2017/06/21 19:27:40 (GMT -400)
  config file = None
  configured module search path = [u'/home/jtanner/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jtanner/workspace/issues/AP-VCSIM_CONTAINER/ansible.provider/lib/ansible
  executable location = /home/jtanner/workspace/issues/AP-VCSIM_CONTAINER/ansible.provider/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]

```

##### SUMMARY
```
Using module file /root/ansible/lib/ansible/modules/cloud/vmware/vmware_guest_find.py
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1498091429.1572943-39737921190087 `" && echo ansible-tmp-1498091429.1572943-39737921190087="` echo /root/.ansible/tmp/ansible-tmp-1498091429
.1572943-39737921190087 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmpy3hvylfc TO /root/.ansible/tmp/ansible-tmp-1498091429.1572943-39737921190087/vmware_guest_find.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1498091429.1572943-39737921190087/ /root/.ansible/tmp/ansible-tmp-1498091429.1572943-39737921190087/vmware_guest_find.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/tmp/ansible-test-coverage-kgav7mhr/coverage/injector.py /root/.ansible/tmp/ansible-tmp-1498091429.1572943-39737921190087/vmware_guest_find.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1498091429.1572943-
39737921190087/" > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_nybjxzyy/ansible_module_vmware_guest_find.py", line 281, in <module>
    main()
  File "/tmp/ansible_nybjxzyy/ansible_module_vmware_guest_find.py", line 267, in main
    uuid=module.params['uuid']
  File "/tmp/ansible_nybjxzyy/ansible_module_vmware_guest_find.py", line 120, in getvm_folder_paths
    folderpath = self.compile_folder_path_for_object(vobj)
  File "/tmp/ansible_nybjxzyy/ansible_module_vmware_guest_find.py", line 203, in compile_folder_path_for_object
    while hasattr(thisobj, 'parent'):
  File "/usr/local/lib/python3.5/dist-packages/pyVmomi/VmomiSupport.py", line 574, in __call__
    return self.f(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/pyVmomi/VmomiSupport.py", line 394, in _InvokeAccessor
    return self._stub.InvokeAccessor(self, info)
  File "/usr/local/lib/python3.5/dist-packages/pyVmomi/StubAdapterAccessorImpl.py", line 43, in InvokeAccessor
    objectContent = result.objects[0]
IndexError: list index out of range

failed: [testhost] (item=/F0/DC0/vm/F0/DC0_H0_VM1) => {
    "changed": false,
    "failed": true,
    "item": "/F0/DC0/vm/F0/DC0_H0_VM1",
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_nybjxzyy/ansible_module_vmware_guest_find.py\", line 281, in <module>\n    main()\n  File \"/tmp/ansible_nybjxzyy/ansible_module_vmware_guest_find.py\"
, line 267, in main\n    uuid=module.params['uuid']\n  File \"/tmp/ansible_nybjxzyy/ansible_module_vmware_guest_find.py\", line 120, in getvm_folder_paths\n    folderpath = self.compile_folder_path_for_object(vobj)\n  File \"/tmp/a
nsible_nybjxzyy/ansible_module_vmware_guest_find.py\", line 203, in compile_folder_path_for_object\n    while hasattr(thisobj, 'parent'):\n  File \"/usr/local/lib/python3.5/dist-packages/pyVmomi/VmomiSupport.py\", line 574, in __ca
ll__\n    return self.f(*args, **kwargs)\n  File \"/usr/local/lib/python3.5/dist-packages/pyVmomi/VmomiSupport.py\", line 394, in _InvokeAccessor\n    return self._stub.InvokeAccessor(self, info)\n  File \"/usr/local/lib/python3.5/
dist-packages/pyVmomi/StubAdapterAccessorImpl.py\", line 43, in InvokeAccessor\n    objectContent = result.objects[0]\nIndexError: list index out of range\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}
```

##### STEPS TO REPRODUCE
Run the test from https://github.com/ansible/ansible/pull/25964 with the ubuntu1604py3 container

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
* Remove redundant get_obj method
* Fix testcase
* Correct logic for compile_folder_path_for_object

Fixes: #25984

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/vmware.py
lib/ansible/modules/cloud/vmware/vmware_guest_find.py
test/integration/targets/vmware_guest_find/aliases
test/integration/targets/vmware_guest_find/tasks/main.yml

##### ANSIBLE VERSION
```
2.4devel
```

### CODE REVIEW

**Comment by abadger** [comment]
This looks good to me but all of the important bits depend on how the vmware API works.  I don't have any experience with that.  If one of the maintainers would like to take a look and approve feel free to ping me and I'll merge this.

**Comment by abadger** [comment]
@Akasurde I'd also be willing to merge this if you can just find someone to test it.  Really just another set of eyes since I don't use vmware and don't knwo the API itself.

**Comment by Akasurde** [comment]
@dagwieers @pdellaert @dav1x @jctanner Could you please take a look into this PR ?

File: test/integration/targets/vmware_guest_find/tasks/main.yml
Comment (dagwieers) [inline test/integration/targets/vmware_guest_find/tasks/main.yml]: We now use a one-line GPL license header.

File: lib/ansible/module_utils/vmware.py
Comment (dagwieers) [inline lib/ansible/module_utils/vmware.py:237]: This is not doing exactly the same thing (maybe on purpose ?).

Originally it would remove 'Datacenters' if it was the first item in paths. Now it seems you're removing everything after the first encountered 'Datacenters'. Can you confirm this change of implementation is exactly what you want ?

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: test/integration/targets/vmware_guest_find/tasks/main.yml
Comment (Akasurde) [reply test/integration/targets/vmware_guest_find/tasks/main.yml]: ok.

File: lib/ansible/module_utils/vmware.py
Comment (pdellaert) [reply lib/ansible/module_utils/vmware.py:237]: Doesn't look like that should be the intention... I guess it should be `continue`?

Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py:237]: @dagwieers @pdellaert 

This while loop is trying to build path for a object. We need to break when we encounter 'Datacenters' which is parent of every object. In Python2*, it works perfectly, but in Python3, it fails. 
So I added `break` when we encounter Datacenters, since we don't have Datacenters in paths list then, there is no point in keeping `if paths[0] == 'Datacenters':` statement.
Without fix [this script](https://github.com/Akasurde/pyvmomi_examples/blob/master/compile_path_example.py) fails
With fix: 
```
# python compile_path_example.py
/
/vm
/datastore
/network
/
# python3 compile_path_example.py
/
/vm
/datastore
/network
/
# govc find /
/
/DC0
/DC0/vm
/DC0/vm/DC0_H0_VM0
/DC0/vm/DC0_H0_VM1
/DC0/vm/DC0_C0_RP0_VM0
/DC0/vm/DC0_C0_RP0_VM1
/DC0/host
/DC0/host/DC0_H0
/DC0/host/DC0_H0/DC0_H0
/DC0/host/DC0_H0/Resources
/DC0/host/DC0_C0
/DC0/host/DC0_C0/DC0_C0_H0
/DC0/host/DC0_C0/DC0_C0_H1
/DC0/host/DC0_C0/DC0_C0_H2
/DC0/host/DC0_C0/Resources
/DC0/datastore
/DC0/datastore/LocalDS_0
/DC0/network
/DC0/network/VM Network
/DC0/network/DVS0
/DC0/network/DC0_DVPG0
```

Comment (pdellaert) [reply lib/ansible/module_utils/vmware.py:237]: Your code breaks the loop, so it doesn't continue. This assumes that `Datacenters` is always the root folder in all cases, and nothing else above it exists. If that's a certainty, your code is good. Otherwise the question becomes how we should handle something above the `Datacenters` folder

Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py:237]: Yes. 'Datacenters' is parent of all object so I don't think so there will be anything above it. If you see code is doing `paths.reverse()` so `Datacenters` will be the last candidate we are looking for. Can anyone confirm this assumption?

**Comment by Akasurde** [comment]
ready_for_review

File: lib/ansible/module_utils/vmware.py
Comment (pdellaert) [reply lib/ansible/module_utils/vmware.py:237]: As far as i know, you correct in those assumptions. Just worried what will happen if there is actually a subfolder with name 'Datacenters' (for instance: Datacenters/DC01/vm/IT-loads/Datacenters/London/)

Still approving, as it is a corner case

**Review by pdellaert — APPROVED** [review]

**Comment by abadger** [comment]
@akasurde Cool, it looks like the code questions have been resolved.  If you want to change the license headers to one-line (per dag's comment) and rebase I'll get this merged.

**Comment by Akasurde** [comment]
@abadger Done.

**Comment by abadger** [comment]
Merged to devel for inclusion in 2.4.0

**Comment by Akasurde** [comment]
@dagwieers @pdellaert @abadger Thanks for review and merge.

**Pull request merged**
